twitter_api/main.py

```diff
@@ -1,23 +1,315 @@
+    else:
+        abort(404)
+    
+def _get_existing_tweet(tweet_id):
+    # Obtain existing data from tweet DB
+    query = '''SELECT id, user_id, created, content
+               FROM tweet
+               WHERE id=?;'''
+    cursor = g.db.execute(query, (tweet_id,))
+    results = cursor.fetchone()
+    if results:
+        return results
+    else:
+        abort(404)
+    
+@auth_only
+def _delete_existing_tweet(tweet_id, user_id):
+    # Obtain existing data from tweet DB
+    query = '''SELECT id, user_id
+               FROM tweet
+               WHERE id=? AND user_id=?;'''
+    cursor = g.db.execute(query, (int(tweet_id), (int(user_id))))
+    results = cursor.fetchone()
+    if results:
+        query = '''DELETE FROM tweet WHERE id=?;'''
+        g.db.execute(query, (int(tweet_id),))
+        g.db.commit()
+        return True
+    return False
+
+def _get_username_from_userid(user_id):
+    query = '''SELECT username
+               FROM user
+               WHERE id=?;'''
+    cursor = g.db.execute(query, (user_id,))
+    return cursor.fetchone()[0]
+
+
+#############################
+# Views:                    #
+#############################
+@app.route('/', methods=['GET'])
+def base():
+    return Response(status=204)
 
-# implement your views here
+@app.route('/login', methods=['POST'])
+def login():
+    
+    # Verify the correct data was submitted
+    if any(val not in request.json for val in {'username', 'password'}):
+        abort(400)
+    
+    # Get submitted data for authentication
+    un = request.json['username']
+    pw = _hash_password(request.json['password'])
+    
+    # Verify that a user exists who matches the submitted username
+    query = 'SELECT COUNT(*) FROM user WHERE username=?;'
+    count = g.db.execute(query, (un,)).fetchone()[0]
+    if count == 0:
+        abort(404)
+    
+    # Find the user id of the user whose data match the submitted data
+    query = 'SELECT id FROM user WHERE username=? AND password=?;'
+    cursor = g.db.execute(query, (un, pw))
+    results = cursor.fetchone()
+    
+    # If there is an entry in the query result, store it as the user id
+    if results:
+        uid = results[0]
+    
+    # If the user fails authentication, return 401
+    else:
+        abort(401)
+    
+    # Generate an access token
+    ACCESS_TOKEN = _generate_access_token()
+    
+    # Store access token in *auth* table
+    query = 'INSERT INTO "auth" ("user_id", "access_token") VALUES (?, ?);'
+    g.db.execute(query, (int(uid), ACCESS_TOKEN))
+    g.db.commit()
+    response = Response(
+        json.dumps({'access_token': ACCESS_TOKEN}),
+        status=201,
+        mimetype=JSON_MIME_TYPE
+    )
+    
+    return response
+    
 
+@app.route('/logout', methods=['POST'])
+@auth_only
+def logout():
+    '''
+    Logs out the user by removing the user's access token from the auth table
+    '''
+    
+    # Remove the access token
+    query = 'DELETE FROM auth WHERE access_token=?;'
```

Comment:
I could login several times on your service and you'd generate several access_tokens for me. Wouldn't it be better to delete ALL the access tokens tied to the user?